lib/Support/BuiltInFunctions.cpp

```
@@ -1035,3 +1037,24 @@ RUNTIME_FUNC_DEF(zeros, void, ARRAY(int32_t))
 RUNTIME_FUNC_DEF(zeros, void, ARRAY(int64_t))
 RUNTIME_FUNC_DEF(zeros, void, ARRAY(float))
 RUNTIME_FUNC_DEF(zeros, void, ARRAY(double))
+
+//===----------------------------------------------------------------------===//
+// assert
+//===----------------------------------------------------------------------===//
+
+namespace {
+void assert_void(bool condition, void *msg, int64_t level) {
+  std::string inMsg((char *)msg);
+  std::string abortMsg;
+
+  abortMsg = (level) ? "Error: " : "Warning: ";
```

Explicitly check for equality of level with `1` or `0`.

---

Since we are also supporting user-written assertions, can we rely on the fact that we'll never get negative log levels from lowered code or do we need to raise an exception?

---

It would be better to assert that the `level` has one of the expected values. However I'm not sure to understand what you mean. The `level` is coming from the code generation process. If an unexpected value is found, then the compiler is at fault, not the user.

---

Ok thanks. My point was: when the source program contains an assert, anything which is not `AssertionLevel.WARNING` or `AssertionLevel.ERROR` will make compilation fail, while as for assertions inserted by runtime verification during lowering, they have a fixed `level` of 2. So as of now it is impossible to receive a negative `level`, but better to be 100% sure, why not.

---

The value of `level` may become invalid also if the compiler emits function calls that happen to be inconsistent with the runtime APIs. ABI constraints are indeed not represented yet in MLIR, and the fact that everything behaves correctly when linked together is a very (but not entirely) lucky case. One more assertion in a debug build is nothing but beneficial for such simple yet critical things ;). And beside the assertion, as I was saying, I would also explicitly enumerate the values even for a Release build.

---

I wasn't thinking about ABI issues, thank you so much!